docs/extensions/vss.md

```diff
@@ -0,0 +1,82 @@
+CREATE INDEX my_hnsw_index ON my_vector_table USING HNSW (vec);
+```
+
+The index will then be used to accelerate queries that use a `ORDER BY` clause evaluating one of the supported distance metric functions against the indexed columns and a constant vector, followed by a `LIMIT` clause. For example:
+```sql
+SELECT * FROM my_vector_table ORDER BY array_distance(vec, [1,2,3]::FLOAT[3]) LIMIT 3;
+
+# We can verify that the index is being used by checking the EXPLAIN output 
+# and looking for the HNSW_INDEX_SCAN node in the plan
+
+EXPLAIN SELECT * FROM my_vector_table ORDER BY array_distance(vec, [1,2,3]::FLOAT[3]) LIMIT 3;
+
+┌───────────────────────────┐
+│         PROJECTION        │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│             #0            │
+└─────────────┬─────────────┘                             
+┌─────────────┴─────────────┐
+│         PROJECTION        │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│            vec            │
+│array_distance(vec, [1.0, 2│
+│         .0, 3.0])         │
+└─────────────┬─────────────┘                             
+┌─────────────┴─────────────┐
+│      HNSW_INDEX_SCAN      │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│   t1 (HNSW INDEX SCAN :   │
+│           my_idx)         │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│            vec            │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│           EC: 3           │
+└───────────────────────────┘               
+```
+
+By default the HNSW index will be created using the euclidean distance `l2sq` (L2-norm squared) metric, matching DuckDBs `array_distance` function, but other distance metrics can be used by specifying the `metric` option during index creation. For example:
+```sql
+CREATE INDEX my_hnsw_cosine_index ON my_vector_table USING HNSW (vec) WITH (metric = 'cosine');
+```
+
+The following table shows the supported distance metrics and their corresponding DuckDB functions
+
+| Description | Metric | Function |
+| --- | --- | --- |
+| Euclidean distance | `l2sq` | `array_distance` |
+| Cosine similarity | `cosine` | `array_cosine_similarity` |
+| Inner product | `ip` | `array_inner_product` |
+
+## Inserts, Updates, Deletes and Re-Compaction
+
+The HNSW index does support inserting, updating and deleting rows from the table after index creation. However, there are two things to keep in mind:  
+- Its faster to create the index after the table has been populated with data as the initial bulk load can make better use of parallelism on large tables.
+- Deletes are not immediately reflected in the index, but are instead "marked" as deleted, which can cause the index to grow stale over time and negatively impact query quality and performance.
+
+To remedy the last point, you can call the `PRAGMA hnsw_compact_index('<index name>')` pragma function to trigger a re-compaction of the index pruning deleted items, or re-create the index after a significant number of updates.
+
+## Limitations 
+
+- Only vectors consisting of `FLOAT`s are supported at the moment.
```

Comment:
Might be worth mentioning what size DuckDB's floats are? Could also link to the docs. FP32 I believe?
https://duckdb.org/docs/sql/data_types/numeric#floating-point-types